big_tests/tests/mongoose_helper.erl

```diff
@@ -266,22 +266,58 @@ logout_user(Config, User) ->
             end
     end.
 
+% @doc Waits for `Fun` to return `ExpectedValue`.
+% Each time Different value is returned or
+% functions throws an error, we sleep.
+% Sleep time is 2x longer each try
+% but not longer than `max_time`
+factor_backoff(Fun, ExpectedValue, #{attempts := Attempts,
+                                     min_time := Mintime,
+                                     max_time := Maxtime}) ->
+    do_wait_until(Fun, ExpectedValue, #{attempts => Attempts,
+                                        sleep_time => Mintime,
+                                        history => [],
+                                        next_sleep_time_fun => fun(E) ->
+                                                                  min(E * 2, Maxtime)
+                                                          end}).
+
 wait_until(Predicate, Attempts, Sleeptime) ->
-     wait_until(Predicate, true, Attempts, Sleeptime).
-
- wait_until(Fun, ExpectedValue, Attempts, SleepTime) ->
-     wait_until(Fun, ExpectedValue, Attempts, SleepTime, []).
-
- wait_until(_Fun, _ExpectedValue, 0, _SleepTime, History) ->
-     error({badmatch, History});
- wait_until(Fun, ExpectedValue, AttemptsLeft, SleepTime, History) ->
-     case Fun() of
-         ExpectedValue ->
-             ok;
-         OtherValue ->
-             timer:sleep(SleepTime),
-             wait_until(Fun, ExpectedValue, dec_attempts(AttemptsLeft), SleepTime, [OtherValue | History])
-     end.
+    wait_until(Predicate, true, Attempts, Sleeptime).
+
+wait_until(Fun, ExpectedValue, Attempts, SleepTime) ->
+    do_wait_until(Fun, ExpectedValue, #{attempts => Attempts,
+                                        sleep_time => SleepTime,
+                                        history => [],
+                                        next_sleep_time_fun => fun(E) -> E end}).
+
+
+do_wait_until(_Fun, _ExpectedValue, #{attempts := 0, history := History}) ->
+    error({badmatch, History});
+do_wait_until(Fun, ExpectedValue, #{attempts := AttemptsLeft,
+                                    sleep_time := SleepTime,
+                                    history := History,
+                                    next_sleep_time_fun := NextSleepTimeFun} = State) ->
+    try Fun() of
+        ExpectedValue ->
+            ok;
+        OtherValue ->
+            wait_and_continue(Fun, ExpectedValue, OtherValue, State)
+    catch Error:Reason ->
```

Comment:
I think it'd be good to log the `Error` and `Reason` (unless they are already logged somewhere).

Reply:
The current approach is that if ‚wait_until’ eventually fails, we log the history of errors (or different than expected values returned). I think there is no point in logging values that just haven’t changed yet - that’s what this function is for - waiting for those values to change.

Reply:
All right, that's fine.